scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja

```diff
@@ -2476,4 +2477,302 @@ psa_status_t psa_driver_wrapper_asymmetric_decrypt(
     }
 }
 
+psa_status_t psa_driver_wrapper_pake_setup(
+    psa_pake_operation_t *operation,
+    const psa_pake_cipher_suite_t *cipher_suite )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    /* Try setup on accelerators first */
+#if defined(PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT)
+#if defined(PSA_CRYPTO_DRIVER_TEST)
+    status = mbedtls_test_transparent_pake_setup(
+                        &operation->ctx.transparent_test_driver_ctx,
+                        (const libtestdriver1_psa_pake_cipher_suite_t*) cipher_suite );
+    if( status == PSA_SUCCESS )
+        operation->id = MBEDTLS_TEST_TRANSPARENT_DRIVER_ID;
+
+    if( status != PSA_ERROR_NOT_SUPPORTED )
+        return( status );
+#endif /* PSA_CRYPTO_DRIVER_TEST */
+#endif /* PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT */
+
+    /* If software fallback is compiled in, try fallback */
+#if defined(MBEDTLS_PSA_BUILTIN_PAKE)
+            status = mbedtls_psa_pake_setup( operation, cipher_suite );
+            if( status == PSA_SUCCESS )
+                operation->id = PSA_CRYPTO_MBED_TLS_DRIVER_ID;
+            return status;
+#endif /* MBEDTLS_PSA_BUILTIN_PAKE */
+
+    /* Add cases for opaque driver here */
+#if defined(PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT)
+#if defined(PSA_CRYPTO_DRIVER_TEST)
+
+    status = mbedtls_test_opaque_pake_setup(
+                        &operation->ctx.opaque_test_driver_ctx,
+                        (const libtestdriver1_psa_pake_cipher_suite_t*) cipher_suite );
+    if( status == PSA_SUCCESS )
+        operation->id = MBEDTLS_TEST_OPAQUE_DRIVER_ID;
+
+    return( status );
+
+#endif /* PSA_CRYPTO_DRIVER_TEST */
+#endif /* PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT */
+
+    /* Nothing left to try if we fall through here */
+    (void) status;
+    (void) operation;
+    (void) cipher_suite;
+    return( PSA_ERROR_NOT_SUPPORTED );
+}
+
+psa_status_t psa_driver_wrapper_pake_set_password_key(
+    psa_pake_operation_t *operation,
+    uint8_t *key_buffer,
+    size_t key_size )
+{
+    switch( operation->id )
```

Comment:
For an API function that takes a key ID as argument, the generic code looks up the key metadata and key data in the slot, and passes that to the driver wrapper. The driver wrapper dispatches based on the location (which is encoded in the lifetime), and when the location is `LOCAL_STORAGE` (transparent key), it dispatches based on the key type and algorithm. For a setup function of a multipart operation, the driver wrapper records the choice of driver in `operation->id`.

For an API function that continues a multipart operation and does not receive a key ID as argument, the driver wrapper looks up `operation->id` to decide which driver to dispatch to.

The discussion above is missing the case of an API function that continues a multipart operation, but also receives a key ID as an argument. This requires additional design work in the driver interface.

* If this happens because the API starts doing computation before it knows the key, as happens with the key derivation interface, the solution we're going is to buffer arguments in memory and only start doing computation once the key is available.
* If this happens because there are multiple keys, there's an inherent contradiction: with opaque keys, the operation has to be handled by the secure element that contains the key, but what if the keys are in different secure elements? Then the driver interface needs to break down the operation into parts that can be carried out independently. This is very specific to the operation. For key derivation, there is one secret input and some non-secret inputs; the driver is chosen based on the secret input, and the part of the operation that handles non-secret inputs is just an export.

I haven't looked at how this should be done for ECJPAKE.

Reply:
There's only one key for EC J-PAKE, and the computation never starts before we know the key. The only thing that happens before the know the key is setting some parameters:

Reply:
Thanks for explanation.
So I need to re-implement only `set_password_key` to dispatch based on the location.

Reply:
Re-implemented `set_password_key` dispatch to dispatch by `lifetime`.